tests/e2e/chain/validator.go

```diff
@@ -0,0 +1,329 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return stakingtypes.NewMsgCreateValidator(
+		sdk.ValAddress(v.keyInfo.GetAddress()),
+		valPubKey,
+		amount,
+		description,
+		commissionRates,
+		minSelfDelegation,
+	)
+}
+
+func (v *internalValidator) createConfig() error {
+	p := path.Join(v.configDir(), "config")
+	return os.MkdirAll(p, 0o750)
+}
+
+func (v *internalValidator) createNodeKey() error {
+	serverCtx := server.NewDefaultContext()
+	config := serverCtx.Config
+
+	config.SetRoot(v.configDir())
+	config.Moniker = v.moniker
+
+	nodeKey, err := p2p.LoadOrGenNodeKey(config.NodeKeyFile())
+	if err != nil {
+		return err
+	}
+
+	v.nodeKey = *nodeKey
+	return nil
+}
+
+func (v *internalValidator) createConsensusKey() error {
+	serverCtx := server.NewDefaultContext()
+	config := serverCtx.Config
+
+	config.SetRoot(v.configDir())
+	config.Moniker = v.moniker
+
+	pvKeyFile := config.PrivValidatorKeyFile()
+	if err := tmos.EnsureDir(filepath.Dir(pvKeyFile), 0o777); err != nil {
+		return err
+	}
+
+	pvStateFile := config.PrivValidatorStateFile()
+	if err := tmos.EnsureDir(filepath.Dir(pvStateFile), 0o777); err != nil {
+		return err
+	}
+
+	filePV := privval.LoadOrGenFilePV(pvKeyFile, pvStateFile)
+	v.consensusKey = filePV.Key
+
+	return nil
+}
+
+func (v *internalValidator) createKeyFromMnemonic(name, mnemonic string) error {
+	kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, v.configDir(), nil, evmoskr.Option())
+	if err != nil {
+		return err
+	}
+
+	keyringAlgos, _ := kb.SupportedAlgorithms()
+	algo, err := keyring.NewSigningAlgoFromString("eth_secp256k1", keyringAlgos)
+	if err != nil {
+		return err
+	}
+
+	info, err := kb.NewAccount(name, mnemonic, "", sdk.FullFundraiserPath, algo)
+	if err != nil {
+		return err
+	}
+
+	privKeyArmor, err := kb.ExportPrivKeyArmor(name, keyringPassphrase)
+	if err != nil {
+		return err
+	}
+
+	privKey, _, err := sdkcrypto.UnarmorDecryptPrivKey(privKeyArmor, keyringPassphrase)
+	if err != nil {
+		return err
+	}
+
+	v.keyInfo = info
+	v.mnemonic = mnemonic
+	v.privateKey = privKey
+
+	return nil
+}
+
+func (v *internalValidator) createKey(name string) error {
+	mnemonic, err := createMnemonic()
+	if err != nil {
+		return err
+	}
+
+	return v.createKeyFromMnemonic(name, mnemonic)
+}
+
+func (v *internalValidator) export() *Validator {
+	return &Validator{
+		Name:          v.instanceName(),
+		ConfigDir:     v.configDir(),
+		Index:         v.index,
+		Mnemonic:      v.mnemonic,
+		PublicAddress: v.keyInfo.GetAddress().String(),
+	}
+}
+
+func (v *internalValidator) getNodeKey() *p2p.NodeKey {
+	return &v.nodeKey
+}
+
+func (v *internalValidator) getGenesisDoc() (*tmtypes.GenesisDoc, error) {
+	serverCtx := server.NewDefaultContext()
+	config := serverCtx.Config
+	config.SetRoot(v.configDir())
+
+	genFile := config.GenesisFile()
+	doc := &tmtypes.GenesisDoc{}
+
+	if _, err := os.Stat(genFile); err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
```

Comment:
the `else` case of this clause returns an empty genesis doc (from L193). Is this expected?